modin/engines/base/io/text/csv_dispatcher.py

```diff
@@ -55,13 +59,33 @@ def _read(cls, filepath_or_buffer, **kwargs):
         if chunksize is not None:
             return cls.single_worker_read(filepath_or_buffer, **kwargs)
 
-        skiprows = kwargs.get("skiprows")
-        if skiprows is not None and not isinstance(skiprows, int):
+        # header can be skipped because header information is got before
+        # data file partitioning
+        names = kwargs.get("names", None)
+        header_size = cls._define_header_size(
+            kwargs.get("header", "infer"),
+            names,
+        )
+
+        skiprows, monotonic_skiprows, pre_reading = cls._manage_skiprows_parameter(
+            kwargs.get("skiprows", None), header_size
+        )
+        should_handle_skiprows = skiprows is not None and not isinstance(skiprows, int)
+
+        empty_pd_df = pandas.read_csv(
+            filepath_or_buffer, **dict(kwargs, nrows=0, skipfooter=0)
+        )
+        column_names = empty_pd_df.columns
+        # In this case parallel implementation inefficient
+        # TODO: improve performance of text_file_dispatcher._read_rows function
+        if len(column_names) < 32 and should_handle_skiprows and not monotonic_skiprows:
             return cls.single_worker_read(filepath_or_buffer, **kwargs)
+
```

Comment:
is it inefficient only when `len(column_names) < 32`, and if it is, then why?

Reply:
It is inefficient mostly when number of columns is low and number of rows is big enough (number of rows didn't counted since if there are low number of rows and columns both - this file will be small and distribution will not be effective anyway).
In the case of "narrow" files most of the time is spent in the `_read_rows` function which implemented in Python (at worst case function iterating over all rows from file), while if we fallback to pandas implementation, optimized pandas Cython parser can be used.

Reply:
I see. Is there an ability to replace all our logic about rows counting by some fast Cython parser in the future?

Reply:
Yes, i think there are two options we can use. The first is using already existent pandas C parser [function](https://github.com/pandas-dev/pandas/blob/e7ac30df29a6cf4550e6d6ce5b190a6789270fef/pandas/_libs/parsers.pyx#L749) (hope we will able to adopt it's logic for modin needs) or write modin _read_rows in Cython (in this case we will need to add additional Cython dependencies i think).

Reply:
I think for now it's okay to keep parsing logic written in python since we're satisfied by the current performance (do we?). 

Replacing the parsing mechanism will force us to rewrite a lot of logic in the IO part. We should do some research first, about what kind of speedup it can give us (write a small example of `read_csv` with rows counting). Let's create an issue about that.

Reply:
I think it is okay for now.
Yes, i agree, issue was created https://github.com/modin-project/modin/issues/2734.